blog/2023-03-21-concurrency-in-rest-apis/index.mdx

```diff
@@ -0,0 +1,175 @@
+
+### Saving a document
+
+We designed saving a document so that it will use an idempotent `PUT` HTTP method. This method will be used to create a new document, as well as to update an existing one.
+Clients are allowed to set the `documentId`.
+
+The following is a sequence diagram of how the API operation was implemented:
+
+```mermaid
+flowchart TD
+    A([Start]) --> B{Request body present?}
+    B -- Yes --> C[Get documentID path param]
+    C --> D[Get document from DB by documentID]
+    D --> E{Document exists?}
+    E -- No --> F[Save new document]
+    F --> G([HTTP 204])
+    E -- Yes --> H[Get If-Match header value from request]
+    H --> I{Value missing?}
+    I -- Yes --> J([HTTP 428 - Precondition required])
+    I -- No --> K{Request ETag == document Etag?}
+    K -- Yes --> L[Update document]
+    L --> G
+    K -- No -->  N([HTTP 412 -Precondition failed])
+    B -- No ----> O([HTTP 400 - Bad request])
+```
+
+:::tip
+Requirements for this sequence are described in the [RFC 7232](https://www.rfc-editor.org/rfc/rfc7232#section-5)
+:::
+
+### Getting a document
+
+Getting a document is far simpler. Using standard `GET` HTTP method, we will return the document with the provided `documentId`. If there's no document with the provided `documentId`,
+we will return `404 Not Found`.
+
+The following is a sequence diagram of how the API operation was implemented:
+
+```mermaid
+flowchart TD
+    A([Start]) --> B[Resolve document by provided documentID]
+    B --> C{Document exists?}
+    C -- No --> D([HTTP 404])
+    C -- Yes --> E([HTTP 200])
+```
+
+## Sample code
+
+On our <u>[GitHub](https://github.com/gran-software-solutions/code-samples/tree/main/concurrency-in-rest-apis)</u> we've published a sample project that implements the described API.
+The sample code is written in <u>[Kotlin](https://kotlinlang.org/)</u> using <u>[Vert.x](https://vertx.io/)</u> toolkit. It requires having <u>[JDK 11](https://adoptopenjdk.net/)</u> or newer installed
+
+Here's how you can run it:
+
+* Clone the repository
+  ```shell
+  git clone git@github.com:gran-software-solutions/code-samples.git
+  ```
+* Run the following command in the root of the project:
+  ```shell
+  cd concurrency-in-rest-apis
+  ./gradlew run
+  ```
+* This will start the server on port `8888`
+* Open another shell to run the commands to test the API
+* Create a new document with id `7051d73e-9e82-44ca-a0d9-eb8af23b9cf6`
+   ```shell
+   curl -i -X PUT \
+      -H "Content-Type: text/plain" \
+      -d 'some content' \
+      http://localhost:8888/documents/7051d73e-9e82-44ca-a0d9-eb8af23b9cf6
+   ```
+* Get the document with id `7051d73e-9e82-44ca-a0d9-eb8af23b9cf6`
+  ```shell
+  curl -i \
+    http://localhost:8888/documents/7051d73e-9e82-44ca-a0d9-eb8af23b9cf6
+  ```
+* In the response, you will see the `ETag` header. It will look something like this: `ETag: W/"XXXYYYZZZ"`. Copy the value of the header.
```

Comment:
this is the first time when its mentioned that ETag is actually an http header. I think it would be better to explain that in the very beginning